### ISSUE TITLE
add custom uncompressed size header to `pyodide.asm.wasm` to track loading

### ISSUE THREAD

**Issue by fcrozatier** [pyodide/pyodide#2947]
## Problem

There is yet no official way to track pyodide loading but as we discussed how to do this in [https://github.com/pyodide/pyodide/discussions/2927] it turned out that we need to know in advance the uncompressed size of `pyodide.asm.wasm`.

The solution discussed relies on creating a Stream and comparing processed chunks with the contentLength. But the ContentLength header only grabs the compressed file size, and in the end the processed chunks add up to more than that. 

We can use magic numbers to accomodate this, but:
- it's ugly
- the uncompressed file size can change in future release
- there is also the possibility depending on the way the server is configured that not every user gets the same compression, some may get a brotli compressed file, others with uncompatible browsers may get a gzip file etc. This messes-up with the magic number solution.

## Proposal

The clean way to go would be to have a custom header on the file allowing to programmatically know `pyodide.asm.wasm` uncompressed size.

We believe that any feedback we can give users on the loading status of pyodide is valuable.

Would this be possible ?

**Comment by rth** [comment]
The issue is fetching `pyodide.asm.wasm` is only a small fraction of the overall load time. 

For instance, the [REPL](https://pyodide.org/en/stable/console.html)  loads in 1.8s for me in Firefox while the `pyodide.asm.wasm` timing are,

![image](https://user-images.githubusercontent.com/630936/183361441-6f986994-4c65-4cce-9b52-d674bee6f1cd.png)

So that's not going to be very helpful as a load time indicator.

**Comment by fcrozatier** [comment]
I agree that fetching the file is only one part of the process, but it is still a big part and we can track the progress at this stage. I don't know how to track the progress of the compile step but we can still await it.

In practice here is what I have when throttling down to 4G. The progress bar is tracking the fetch:

https://user-images.githubusercontent.com/48696601/183383876-2a0f1e1d-852e-4928-96eb-f29d2b438ade.mov

See the user experience on the left ? I believe what's not helpful is to have no feedback at all.

**Comment by rth** [comment]
OK, it's indeed more useful with slow download speeds where the overall time would be dominated by fetching files and two large files  `pyodide.asm.wasm` and `pyodide.asm.data` (or roughly the same compressed size) are being downloaded in parallel.

So if we wanted to implement a custom header one would need to pass it to the `aws s3` command [here](https://cdn.jsdelivr.net/pyodide/v0.20.0/full/pyodide.asm.data) (and have some python or bash script calculating it)

I'm not sure it's really worth it though. For a given content-encoding (e.g. gzip) the compression ratio for .wasm would be rather similar I think. For instance, in 0.21 there is a x2.9 gzip compression ratio, and it was x3.3 in 0.18 -> that's a 13% variation.  So you could just fix a x3 ratio and use that. I would argue this difference doesn't really matter because this progress bar is not going to be very accurate anyway, and also since it's UI, people will likely not mind if it's off by a bit.

**Comment by hoodmane** [comment]
So you can do:
```js
loadingBarProgress = math.min((loaded) /(compressedSize * 3), 1);
if(done || loadingBarProgress === 1){
	switchToCompilerSpinner();
}
```
It's not clear to me that it's possible to do this without editing the Pyodide loading code, so maybe we could add a feature that calls event listeners for loading progress and for compiling.

**Comment by fcrozatier** [comment]
>  So you could just fix a x3 ratio and use that.

Indeed this is the current solution and what I called in the first post a magic number, meaning it's a hard-coded approximate number that makes it work. In the progress bar demo above I used a 2.9 ratio.

>  I would argue this difference doesn't really matter because this progress bar is not going to be very accurate anyway, and also since it's UI, people will likely not mind if it's off by a bit

Totally agree on this one but the other three points motivating this issue are still relevant:
- it's ugly
- the uncompressed file size can change in future release *(like not by a bit)*
- there is also the possibility depending on the way the server is configured that not every user gets the same compression, some may get a brotli compressed file, others with uncompatible browsers may get a gzip file etc. This messes-up with the magic number solution

I think it's a burden to have to check every so often if the x2.9 ratio is still good enough.
Points 2 and 3 focus on how to be future-proof. This is why I propose a custom header, it's a simple solution.

**Comment by rth** [comment]
> there is also the possibility depending on the way the server is configured that not every user gets the same compression, some may get a brotli compressed file, others with uncompatible browsers may get a gzip file etc. This messes-up with the magic number solution

Yes, you would have to look at the `Content-Encoding` header of the result and adjust the compression ratio based on that. The only possibilities for the JsDelivr CDN are no-compression, gzip or brotli as far as I know. So that's 3 constants.

> it's ugly

I would argue that custom headers are also not very pretty. There are people doing mirror deployments of Pyodide and having this feature work or not work depending on the deployment is not ideal.

Also I checked in more detail, and apparently we actually cannot add a custom header without writing some custom Lamda functions for AWS S3 (and paying for the execution of those)  https://stackoverflow.com/a/43282856/1791279. On the cost-benefit, in terms of added complexity, I'm -1 for this sorry.

**Comment by fcrozatier** [comment]
> apparently we actually cannot add a custom header without writing some custom Lamda functions for AWS S3 (and paying for the execution of those) https://stackoverflow.com/a/43282856/1791279. On the cost-benefit, in terms of added complexity, I'm -1 for this sorry.

Sure ! Setting up a Lambda function for this is a no-go. But it is only needed if you want total control of the header name. 

You can still use the `x-amz-meta-(something)` namespaced headers. What about this solution ? It's not that bad, it can be setup manually on single files in the S3 bucket and a CI script could adjust the value after a new version deploy. 

There could be a `x-amz-meta-uncompressed-length` header as part of the api for the few relevant files.

**Comment by rth** [comment]
> You can still use the x-amz-meta-(something) namespaced headers.

If you want to experiment with it, feel free to make a PR adding this attribute [here](https://cdn.jsdelivr.net/pyodide/v0.20.0/full/pyodide.asm.data). We could test it on dev deployment.

 The current infrastructure is,
 Pyodide S3 -> JSdelivr S3 -> JSdelivr CDN Infra -> [Final CND provider](https://www.jsdelivr.com/network) (CloudFlare, FastLy, QUANTIL  Bunny depending on the location).
 so I'm not sure it would work.

**Comment by fcrozatier** [comment]
I see your point, the infrastructure is more complex than I thought with different CDN depending on location. Not sure the cost-benefit is worth it indeed.

I'll keep hacking around for now. 

Thanks for the explanations.
